controllers/storagecluster/routes.go

```diff
@@ -0,0 +1,154 @@
+	if err != nil {
+		return err
+	}
+	err = r.createCephRGWRoutes(ocsCephRoutes, instance)
+	if err != nil {
+		r.Log.Error(err, "could not create Routes")
+		return err
+	}
+
+	return nil
+}
+
+// ensureDeleted deletes the CephObjectStores owned by the StorageCluster
+func (obj *ocsCephRGWRoutes) ensureDeleted(r *StorageClusterReconciler, sc *ocsv1.StorageCluster) error {
+	foundRoute := &routev1.Route{}
+	routes, err := r.newCephRGWRoutes(sc)
+	if err != nil {
+		return err
+	}
+
+	for _, route := range routes {
+		err := r.Client.Get(context.TODO(), types.NamespacedName{Name: route.Name, Namespace: sc.Namespace}, foundRoute)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				r.Log.Info("Uninstall: Route not found", "Route Name", route.Name)
+				continue
+			}
+			return fmt.Errorf("Uninstall: Unable to retrieve route %v: %v", route.Name, err)
+		}
+
+		if route.GetDeletionTimestamp().IsZero() {
+			r.Log.Info("Uninstall: Deleting route", "Route Name", route.Name)
+			err = r.Client.Delete(context.TODO(), foundRoute)
+			if err != nil {
+				return fmt.Errorf("Uninstall: Failed to delete route %v: %v", route.Name, err)
+			}
+		}
+
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: route.Name, Namespace: sc.Namespace}, foundRoute)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				r.Log.Info("Uninstall: Route is deleted", "Route Name", route.Name)
+				continue
+			}
+		}
+		return fmt.Errorf("Uninstall: Waiting for route %v to be deleted", route.Name)
+
+	}
+	return nil
+}
+
+// createCephObjectStore creates CephObjectStore in the desired state
+func (r *StorageClusterReconciler) createCephRGWRoutes(routes []*routev1.Route, instance *ocsv1.StorageCluster) error {
+	for _, route := range routes {
+		existing := routev1.Route{}
+		err := r.Client.Get(context.TODO(), types.NamespacedName{Name: route.Name, Namespace: route.Namespace}, &existing)
+		switch {
+		case err == nil:
+			reconcileStrategy := ReconcileStrategy(instance.Spec.ManagedResources.CephObjectStores.ReconcileStrategy)
+			if reconcileStrategy == ReconcileStrategyInit {
+				return nil
+			}
+			if existing.DeletionTimestamp != nil {
+				err := fmt.Errorf("failed to restore route object %s because it is marked for deletion", existing.Name)
+				r.Log.Info("route restore failed")
+				return err
+			}
+
+			r.Log.Info(fmt.Sprintf("Restoring original route %s", route.Name))
+			existing.ObjectMeta.OwnerReferences = route.ObjectMeta.OwnerReferences
+			route.ObjectMeta = existing.ObjectMeta
+			err = r.Client.Update(context.TODO(), route)
+			if err != nil {
+				r.Log.Error(err, fmt.Sprintf("failed to update route Object: %s", route.Name))
+				return err
+			}
+		case errors.IsNotFound(err):
+			r.Log.Info(fmt.Sprintf("creating route %s", route.Name))
+			err = r.Client.Create(context.TODO(), route)
+			if err != nil {
+				r.Log.Error(err, fmt.Sprintf("failed to create CephObjectStore object: %s", route.Name))
+				return err
+			}
+		}
```

Comment:
If the error is something other that `IsNotFound(err)`, where is that handled? Seems like it should fail the reconcile.

Reply:
+1

Reply:
This is the same behaviour as that of the types being reconciled.  None of them check for any error type.

This would also be a good candidate for a new PR where the behaviour is modified for all types.